### PR TITLE
chore: events: implement event observer deregister method

### PR DESCRIPTION
**Pull request by frrist**
- guarded with test

## Related Issues
Lily schedles jobs that use the events system to observe head changes. When these jobs exit their observer needs to be deregistered from the events system. Currently this is not possible, this PR addresses this. 

## Proposed Changes
Allow event observers to be deregistered

## Checklist

Before you mark the PR ready for review, please make sure that:
- [x] All commits have a clear commit message.
- [x] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [x] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] CI is green

### CODE REVIEW

File: chain/events/observer.go [hunk 1]
```diff
@@ -25,7 +25,9 @@ type observer struct {
 	lk        sync.Mutex
 	head      *types.TipSet
 	maxHeight abi.ChainEpoch
-	observers []TipSetObserver
+
+	observerID uint64
+	observers  map[uint64]TipSetObserver
```
Comment (magik6k) [inline chain/events/observer.go:30]: Some things may care about the order this map in iterated in (mostly tests tho)

Comment (frrist) [reply chain/events/observer.go:30]: I could make a `TipSetObservers` type and implement an iterator pattern over it for consistent ordering, would that work?

**Comment by Stebalien** [comment]
Here's a sketch: https://github.com/filecoin-project/lotus/pull/8445

**Comment by frrist** [comment]
Agreed, its a bit complicated to maintain consistent ordering with this approach. Yeah passing the TipSetObserver interface to an unregister method will work for Lily's use case, thanks for the sketch.

**Review by Stebalien — APPROVED** [review]

**Review by magik6k — APPROVED** [review]

**Pull request merged**